xRemoteFile: Randomly fails to download files from GitHub with error "The remote name could not be resolved"

<!--
    Your feedback and support is greatly appreciated, thanks for contributing!

    ISSUE TITLE:
    Please prefix the issue title with the resource name, e.g.
    'ResourceName: Short description of my issue'

    ISSUE DESCRIPTION (this template):
    Please provide information regarding your issue under each header below.
    Write N/A under any headers that do not apply to your issue, or if the
    information is not available.

    NOTE! Sensitive information should be obfuscated.

    PLEASE KEEP THE HEADERS.

    You may remove this comment block, and the other comment blocks,
    but please keep the headers.
-->
#### Details of the scenario you tried and the problem that is occurring

I use resource xRemoteFile from Azure VMs to download release assets hosted on GitHub.
Regularly (about 1 out of 4 times) the download fails with the error "The remote name could not be resolved".
This is very painful because the DSC deployment is part of an ARM template deployment, and when it occurs the whole template deployment appears in error state in Azure.
Of course the URL is perfectly valid and it works fine on next attempt to download it.

#### Verbose logs showing the problem

```
VERBOSE: [2019-07-02 07:07:41Z] [VERBOSE] [SP]:                            [[xRemoteFile]DownloadLdapcp] 
DestinationPath 'C:\Setup\LDAPCP.wsp' doesn't exist on the machine.
VERBOSE: [2019-07-02 07:07:41Z] [VERBOSE] [SP]: LCM:  [ End    Test     ]  [[xRemoteFile]DownloadLdapcp]  in 0.0310 
seconds.
VERBOSE: [2019-07-02 07:07:41Z] [VERBOSE] [SP]: LCM:  [ Start  Set      ]  [[xRemoteFile]DownloadLdapcp]
VERBOSE: [2019-07-02 07:07:41Z] [VERBOSE] [SP]:                            [[xRemoteFile]DownloadLdapcp] Downloading 
'https://github.com/Yvand/LDAPCP/releases/download/12.0.20190321.770/LDAPCP.wsp' to 'C:\Setup\LDAPCP.wsp'.
VERBOSE: [2019-07-02 07:07:41Z] [VERBOSE] [SP]:                            [[xRemoteFile]DownloadLdapcp] GET 
https://github.com/Yvand/LDAPCP/releases/download/12.0.20190321.770/LDAPCP.wsp with 0-byte payload
VERBOSE: [2019-07-02 07:07:52Z] [VERBOSE] [SP]: LCM:  [ End    Set      ]  [[xRemoteFile]DownloadLdapcp]  in 12.4380 
seconds.
VERBOSE: [2019-07-02 07:07:52Z] [ERROR] PowerShell DSC resource MSFT_xRemoteFile  failed to execute Set-TargetResource 
functionality with error message: Invoking web request failed with error. The remote name could not be resolved: 
'github-production-release-asset-2e65be.s3.amazonaws.com' 
```

#### Suggested solution to the issue

I implemented a dumb retry mechanism directly in the resource to mitigate this error:

```powershell
$nbAttempt = 0
$success = $false
do
{
	try
	{
		$nbAttempt++
		Invoke-WebRequest @PSBoundParameters -Headers $headersHashtable -OutFile $DestinationPath
		$success = $true
	}
	catch [System.Exception]
	{
		Write-Verbose -Message ("YVAND: Download of resource failed on attempt $nbAttempt")
		if ($nbAttempt -gt 5)
		{
			# Inside catch variable $_ is not the exception itself, but System.Management.Automation.ErrorRecord that contains Exception
			throw $_.Exception
		}
	}
}
while ($success -eq $false)
```

I am currently testing it to validate if this workaround is actually working.

#### The DSC configuration that is used to reproduce the issue (as detailed as possible)

```powershell
xRemoteFile DownloadLdapcp
{
	Uri             = "https://github.com/Yvand/LDAPCP/releases/download/12.0.20190321.770/LDAPCP.wsp"
	DestinationPath = "C:\Setup\LDAPCP.wsp"
}
```

#### The operating system the target node is running
<!--
    Please provide as much as possible about the target node, for example
    edition, version, build and language.
    On OS with WMF 5.1 the following command can help get this information.

    Get-ComputerInfo -Property @(
        'OsName',
        'OsOperatingSystemSKU',
        'OSArchitecture',
        'WindowsVersion',
        'WindowsBuildLabEx',
        'OsLanguage',
        'OsMuiLanguages')
-->

```
OsName               : Microsoft Windows Server 2019 Datacenter
OsOperatingSystemSKU : DatacenterServerEdition
OsArchitecture       : 64-bit
WindowsVersion       : 1809
WindowsBuildLabEx    : 17763.1.amd64fre.rs5_release.180914-1434
OsLanguage           : en-US
OsMuiLanguages       : {en-US}
```

#### Version and build of PowerShell the target node is running
<!--
    To help with this information, please run this command:
    $PSVersionTable
-->

```
Name                           Value                                                                  
----                           -----                                                                  
PSVersion                      5.1.17763.316                                                          
PSEdition                      Desktop                                                                
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                
BuildVersion                   10.0.17763.316                                                         
CLRVersion                     4.0.30319.42000                                                        
WSManStackVersion              3.0                                                                    
PSRemotingProtocolVersion      2.3                                                                    
SerializationVersion           1.1.0.1
```

#### Version of the DSC module that was used ('dev' if using current dev branch)

xPSDesiredStateConfiguration 8.8.0.0

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,9 @@
 
 ## Unreleased
 
+- MSFT_xRemoteFile:
+  - Add a retry mechanism when the download fails.
+
 ## 8.8.0.0
 
 - Ports fix for the following issue:

--- a/DSCResources/MSFT_xRemoteFile/MSFT_xRemoteFile.psm1
+++ b/DSCResources/MSFT_xRemoteFile/MSFT_xRemoteFile.psm1
@@ -253,24 +253,30 @@ function Set-TargetResource
         $ProgressPreference = 'SilentlyContinue'
 
         Write-Verbose -Message ($script:localizedData.DownloadingURI -f $DestinationPath, $URI)
-        $nbAttempt = 0
+        $count = 0
         $success = $false
+
         do
         {
             try
             {
-                $nbAttempt++
-                Invoke-WebRequest @PSBoundParameters -Headers $headersHashtable -OutFile $DestinationPath
+                $count++
+                Invoke-WebRequest `
+                    @PSBoundParameters `
+                    -Headers $headersHashtable `
+                    -OutFile $DestinationPath
                 $success = $true
             }
             catch [System.Exception]
             {
-                Write-Verbose -Message ($script:localizedData.DownloadingFailedRetry -f $URI, $nbAttempt, $_.Exception.Message)
-                if ($nbAttempt -gt 5)
+                Write-Verbose -Message ($script:localizedData.DownloadingFailedRetry -f $URI, $count, $_.Exception.Message)
+
+                if ($count -gt 5)
                 {
                     # Inside catch variable $_ is not the exception itself, but a System.Management.Automation.ErrorRecord that contains the actual Exception
                     throw $_.Exception
                 }
+
                 Start-Sleep -Seconds 5
             }
         }
@@ -294,7 +300,6 @@ function Set-TargetResource
     {
         $ProgressPreference = $currentProgressPreference
     }
-
 
     # Update cache
     if (Test-Path -Path $DestinationPath)

--- a/DSCResources/MSFT_xRemoteFile/MSFT_xRemoteFile.psm1
+++ b/DSCResources/MSFT_xRemoteFile/MSFT_xRemoteFile.psm1
@@ -271,7 +271,7 @@ function Set-TargetResource
                     # Inside catch variable $_ is not the exception itself, but a System.Management.Automation.ErrorRecord that contains the actual Exception
                     throw $_.Exception
                 }
-				Start-Sleep -Seconds 5
+                Start-Sleep -Seconds 5
             }
         }
         while ($success -eq $false)

--- a/DSCResources/MSFT_xRemoteFile/MSFT_xRemoteFile.psm1
+++ b/DSCResources/MSFT_xRemoteFile/MSFT_xRemoteFile.psm1
@@ -253,8 +253,28 @@ function Set-TargetResource
         $ProgressPreference = 'SilentlyContinue'
 
         Write-Verbose -Message ($script:localizedData.DownloadingURI -f $DestinationPath, $URI)
-
-        Invoke-WebRequest @PSBoundParameters -Headers $headersHashtable -OutFile $DestinationPath
+        $nbAttempt = 0
+        $success = $false
+        do
+        {
+            try
+            {
+                $nbAttempt++
+                Invoke-WebRequest @PSBoundParameters -Headers $headersHashtable -OutFile $DestinationPath
+                $success = $true
+            }
+            catch [System.Exception]
+            {
+                Write-Verbose -Message ($script:localizedData.DownloadingFailedRetry -f $URI, $nbAttempt, $_.Exception.Message)
+                if ($nbAttempt -gt 5)
+                {
+                    # Inside catch variable $_ is not the exception itself, but a System.Management.Automation.ErrorRecord that contains the actual Exception
+                    throw $_.Exception
+                }
+				Start-Sleep -Seconds 5
+            }
+        }
+        while ($success -eq $false)
     }
     catch [System.OutOfMemoryException]
     {
@@ -274,6 +294,7 @@ function Set-TargetResource
     {
         $ProgressPreference = $currentProgressPreference
     }
+
 
     # Update cache
     if (Test-Path -Path $DestinationPath)

--- a/DSCResources/MSFT_xRemoteFile/en-us/MSFT_xRemoteFile.strings.psd1
+++ b/DSCResources/MSFT_xRemoteFile/en-us/MSFT_xRemoteFile.strings.psd1
@@ -12,6 +12,7 @@ ConvertFrom-StringData @'
     DownloadOutOfMemoryException=Invoking web request failed with OutOfMemoryException- Possible cause is the requested file being too big. {0}
     DownloadException=Invoking web request failed with error. {0}
     DownloadingURI=Downloading '{1}' to '{0}'.
+    DownloadingFailedRetry=Download of '{0}' failed on attempt {1} with this error: {2}
     CacheReflectsCurrentState=Cache reflects current state. No need for downloading file.
     CacheIsEmptyOrNotMatchCurrentState=Cache is empty or it doesn't reflect current state. File will be downloaded.
     MatchSourceFalse=MatchSource is false. No need for downloading file.

